src/wps/core/repository.py

```diff
@@ -192,10 +280,44 @@ async def create(
         )
         await self._dao.insert(work_package)
         encrypted_token = encrypt(token, user_public_crypt4gh_key)
+
         return WorkPackageCreationResponse(
             id=str(work_package.id), token=encrypted_token, expires=expires
         )
 
+    def _validate_work_package(self, work_package: WorkPackage):
+        """Validate certain work package details.
+
+        Raises a WorkPackageAccessError in the following cases:
+        - the work package has expired
+        - the work package is for download and the dataset ID is missing
+        - the work package is for upload and the box ID is missing
+        """
+        extra = {"work_package_id": work_package.id}  # only used for logging
+        if not work_package.created <= now_utc_ms_prec() < work_package.expires:
+            access_error = self.WorkPackageAccessError("Work package has expired")
+            log.error(access_error, extra=extra)
+            raise access_error
+
+        dataset_id = work_package.dataset_id
+        if work_package.type == WorkPackageType.DOWNLOAD and not dataset_id:
```

Comment:
shouldn't we simply check this in a WorkPackage validator, just as we do with the WorkPackageCreationData?